docs(core): add note about not mutating multi provider arrays

crisbeto opened the pull request:
Adds a note to the provider docs that users shouldn't mutate an array that 
is returned from a `multi` provider, because it can cause unforeseen 
consequences in other parts of the app.

Closes #37481.

AndrewKushnir commented on aio/content/guide/dependency-injection-providers.md:
```diff
@@ -278,6 +278,11 @@ When you provide multiple sets of routes using [RouterModule.forRoot](api/router
 and [RouterModule.forChild](api/router/RouterModule#forchild) in a single module,
 the [ROUTES](api/router/ROUTES) token combines all the different provided sets of routes into a single value.
 
+Note that the array that is returned for a `multi` provider is immutable in development mode, which
```
I think we should mention that it's in Ivy only + may be add a note to Ivy compatibility guide?

crisbeto replied:
My concern with mentioning Ivy here is that there's nothing that'll alert us to remove it once Ivy is the only rendering engine option.

AndrewKushnir reviewed (commented):
Thanks for this update @crisbeto 👍 

I believe this might be a risky change to merge into a patch branch. May be we should target "master-only" to have it in 10.1.0?

I also added @pkozlowski-opensource and @mhevery as reviewers to provide additional feedback.

Thank you.

crisbeto commented:
I think what we discussed was to run a presubmit for these changes to get a better idea of how breaking it would be.

AndrewKushnir commented:
Hi @crisbeto, right, I was thinking that it'd be breaking anyways (so we can apply labels, etc), we can just estimate the impact after running presubmit in Google's codebase. I've started it a few mins ago ([link](http://test/OCL:317733969:BASE:317734045:1592860792801:e02b7f91)) and will let you know the result. Thank you.

pkozlowski-opensource reviewed (approved):
LGTM from the code change point of view and based on what we've discussed during the fwk sync. 
A bit concerned that we do this freezing on the one off basis (there are probably other parts of the framework that we should review) but this is outside of the scope of this PR.

Marking as blocked since we need to wait for the presubmit results to see how breaky it is.

crisbeto commented:
@AndrewKushnir I've turned this into a docs change as discussed a while ago. Can you take another look?

AndrewKushnir commented:
Thanks for the update @crisbeto. I have a couple thoughts/ideas:

- should we highlight this section (using `<div class="alert is-helpful">..</div>`)? It feels like it's quite important piece of information.
- would it make sense to add an example on how/where to set the `Readonly` type?

Not related to this PR: it looks like we should consider splitting the `Predefined tokens and multiple providers` into 2 sections: one that talks about multi providers in general (how to use them) and the second section that describes predefined (multi) tokens. Currently it reads like multi providers can only be used to extend the built-in tokens, when in fact multi providers can be used with other (user-defined) tokens. @aikidave can we add that to the docs team backlog?

crisbeto commented:
Good points @AndrewKushnir, I've incorporated both of them. My only concern is that now we have two `<div class="alert is-helpful">` in a row.

mary-poppins commented:
You can preview 212b15a at https://pr37645-212b15a.ngbuilds.io/.
You can preview c550aad at https://pr37645-c550aad.ngbuilds.io/.

mary-poppins commented:
You can preview 67cdbc7 at https://pr37645-67cdbc7.ngbuilds.io/.

AndrewKushnir commented on aio/content/guide/dependency-injection-providers.md:
```diff
@@ -284,6 +284,22 @@ Search for [Constants in API documentation](api?type=const) to find more built-i
+
+</div>
+
+For example, this is how you can type your `multi` provider to avoid unwanted array mutations:
```
```suggestion
You can use `ReadonlyArray` to type your `multi` provider, so TypeScript triggers an error in case of unwanted array mutations:
```

AndrewKushnir reviewed (commented):
Thanks @crisbeto! I've added a couple more comments - feel free to use (or ignore) them as you see fit.

AndrewKushnir commented:
> Not related to this PR: it looks like we should consider splitting the Predefined tokens and multiple providers section (see here) into 2 sections...

FYI, I also found a PR #29909 that introduces the multi provider section (but it looks like it needs some cleanup). cc @aikidave

crisbeto committed:
docs(core): add note about not mutating multi provider arrays

Adds a note to the provider docs that users shouldn't mutate an array that
is returned from a `multi` provider, because it can cause unforeseen
consequences in other parts of the app.

Closes #37481.

crisbeto commented:
I've incorporated the latest set of feedback.

mary-poppins commented:
You can preview d15bac8 at https://pr37645-d15bac8.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Thanks for the changes @crisbeto 👍

pkozlowski-opensource reviewed (approved):
LGTM
